src/datacollector/DataCollectionAttachmentManager.cs

```diff
@@ -285,141 +201,83 @@ private static void Validate(AttachmentRequest attachmentRequest)
-        /// The user token.
-        /// </param>
-        /// <param name="exception">
-        /// The exception.
-        /// </param>
-        /// <param name="path">
-        /// The path.
-        /// </param>
-        private void TriggerCallback(
-            AsyncCompletedEventHandler transferCompletedCallback,
-            object userToken,
-            Exception exception,
-            string path)
-        {
-            Debug.Assert(!string.IsNullOrEmpty(path), "null or empty path");
-            if (transferCompletedCallback != null)
-            {
-                try
-                {
-                    transferCompletedCallback(this, new AsyncCompletedEventArgs(exception, false, userToken));
-                }
-                catch (Exception e)
-                {
-                    if (EqtTrace.IsErrorEnabled)
-                    {
-                        EqtTrace.Error(
-                            "DataCollectionAttachmentManager.TriggerCallBack: Error occurred while raising the file transfer completed callback for {0}. Error: {1}",
-                            path,
-                            e.ToString());
-                    }
-                }
-            }
-        }
-
-        /// <summary>
-        /// Move attachment.
-        /// </summary>
-        /// <param name="attachmentRequest">Attachment request</param>
-        private void MoveAttachment(AttachmentRequest attachmentRequest)
-        {
-            Exception error = null;
-            try
-            {
-                Validate(attachmentRequest);
-                File.Move(attachmentRequest.FileTransferInfo.FileName, attachmentRequest.LocalFilePath);
-            }
-            catch (Exception ex)
-            {
-                error = ex;
-            }
-            finally
-            {
-                // Let collector know of transfer completed if it has registered transferCompletedHandler.
-                this.TriggerCallback(
-                    attachmentRequest.FileTransferInfo.FileTransferCompletedHandler,
-                    attachmentRequest.FileTransferInfo.UserToken,
-                    error,
-                    attachmentRequest.FileTransferInfo.FileName);
-                CompleteFileTransfer(attachmentRequest, error);
-            }
+            var testCaseId = fileTransferInfo.Context.HasTestCase
+                                 ? fileTransferInfo.Context.TestExecId.Id.ToString()
+                                 : string.Empty;
+
+            var directoryPath = Path.Combine(
+                this.SessionOutputDirectory,
+                testCaseId.ToString());
+            var localFilePath = Path.Combine(directoryPath, Path.GetFileName(fileTransferInfo.FileName));
+
+            var task = new Task(() =>
+             {
+                 Validate(fileTransferInfo, localFilePath);
+
+                 try
+                 {
+                     if (fileTransferInfo.PerformCleanup)
```

Comment:
It is possible user may want to cancel a test run or the runner/communication channel crash during these operations. What is the strategy in such cases?